In [1]:
def wordle(guess, answer):
    guess = guess.lower()
    answer = answer.lower()
    result = ['⬜️'] * 5

    
    answer_letters = list(answer)
    
    # handle greens
    for i in range(5):
        if guess[i] == answer[i]:
            result[i] = '🟩'
            answer_letters[i] = None  # remove matched letter
    
    # handle yellows
    for i in range(5):
        if result[i] == '⬜️' and guess[i] in answer_letters:
            result[i] = '🟨'
            answer_letters[answer_letters.index(guess[i])] = None  # mark that letter as used
    
    return ''.join(result)

In [2]:
# uses the same rules as the NYT wordle
wordle(guess = 'AAAAA', answer = 'AABAA')

'🟩🟩⬜️🟩🟩'

In [3]:
import pandas as pd

csv = pd.read_csv('wordle_all_guesses.csv')
csv = csv[['Game Number','Guess','username', 'Answer', 'Output']]
csv 

,Game Number,Guess,username,Answer,Output
0,1456,1,user_0,GHOST,⬜️⬜️⬜️⬜️⬜️
1,1456,2,user_0,GHOST,🟨🟩🟩⬜️⬜️
2,1456,3,user_0,GHOST,🟨🟩🟩🟨⬜️
3,1456,4,user_0,GHOST,🟩🟩🟩🟩🟩
4,1540,1,user_1,BULGE,⬜️🟨⬜️⬜️⬜️
...,...,...,...,...,...
774,1601,2,user_1,GUISE,⬜️⬜️🟨🟩⬜️
775,1601,3,user_1,GUISE,⬜️⬜️🟨🟩🟩
776,1601,4,user_1,GUISE,⬜️🟩⬜️🟩🟩
777,1601,5,user_1,GUISE,⬜️🟨⬜️🟨🟩


In [4]:
csv = csv[csv['Guess'] == 1]
csv = csv[['Game Number','username', 'Answer', 'Output']]
csv

,Game Number,username,Answer,Output
0,1456,user_0,GHOST,⬜️⬜️⬜️⬜️⬜️
4,1540,user_1,BULGE,⬜️🟨⬜️⬜️⬜️
8,1540,user_2,BULGE,⬜️⬜️⬜️🟨⬜️
12,1540,user_0,BULGE,⬜️⬜️⬜️🟨🟨
18,1541,user_0,TENOR,⬜️⬜️⬜️🟨⬜️
...,...,...,...,...
756,1601,user_18,GUISE,⬜️🟩⬜️🟨⬜️
760,1601,user_2,GUISE,⬜️⬜️⬜️⬜️⬜️
764,1601,user_12,GUISE,⬜️🟨🟨🟨⬜️
768,1601,user_3,GUISE,🟨⬜️⬜️⬜️⬜️


In [5]:
# find the unique values in the username col
usernames = csv['username'].unique()
l = list(usernames)
for i, u in enumerate(l):
    print(i,u)

0 user_0
1 user_1
2 user_2
3 user_3
4 user_4
5 user_5
6 user_6
7 user_7
8 user_8
9 user_9
10 user_10
11 user_11
12 user_12
13 user_13
14 user_14
15 user_15
16 user_16
17 user_17
18 user_18
19 user_19
20 user_20
21 user_21


In [6]:
suspect = usernames[4]
suspect

'user_4'

In [7]:
# find the subset of a specific user
subset_suspect = csv[csv['username'] == suspect]
subset_suspect

,Game Number,username,Answer,Output
24,1541,user_4,TENOR,🟨🟩⬜️⬜️⬜️
32,1542,user_4,CHIRP,🟨⬜️⬜️🟨⬜️
44,1543,user_4,TRICK,🟨⬜️⬜️🟨⬜️
46,1544,user_4,POUTY,🟩⬜️🟨🟨⬜️
55,1546,user_4,THROB,🟩🟩⬜️⬜️⬜️
60,1550,user_4,LEFTY,⬜️🟩⬜️🟩🟩
62,1551,user_4,TEETH,🟩🟩⬜️🟨⬜️
64,1552,user_4,KNIFE,⬜️🟨⬜️🟨⬜️
68,1553,user_4,LATER,🟨🟨⬜️⬜️🟨
71,1554,user_4,DEFER,🟩🟩⬜️⬜️⬜️


In [8]:
answers = subset_suspect['Answer']
feedbacks = subset_suspect['Output']

In [9]:
for answer, feedback in zip(answers, feedbacks):    
    output = wordle('REGIA', answer)
    is_correct = output == feedback
    print(f'{answer}, {(not is_correct)* feedback}')


TENOR, 
CHIRP, 
TRICK, 
POUTY, 🟩⬜️🟨🟨⬜️
THROB, 🟩🟩⬜️⬜️⬜️
LEFTY, ⬜️🟩⬜️🟩🟩
TEETH, 🟩🟩⬜️🟨⬜️
KNIFE, 
LATER, 
DEFER, 🟩🟩⬜️⬜️⬜️
QUILL, 🟩🟩🟩⬜️⬜️
MOUTH, 🟨🟩🟩⬜️🟩
BLUNT, 
DRAPE, ⬜️🟩🟩🟩🟩
DALLY, ⬜️🟨🟨⬜️⬜️
FRITZ, 
GOOEY, 
CIVIL, ⬜️⬜️⬜️⬜️⬜️
GEESE, ⬜️⬜️⬜️⬜️⬜️
SPOIL, 🟩⬜️⬜️🟩🟩
WIDTH, 
SPASM, ⬜️⬜️⬜️⬜️⬜️
RELAY, 
PLANE, 🟩🟩🟩⬜️🟨
AMUSE, 
NYLON, 
ANNOY, 
HARDY, 
LEVER, 
STACK, 
WOUND, 
BEARD, 
FORUM, 
SPOOF, 
CATTY, 
GROSS, 
HAVEN, 
IDEAL, 
LIMBO, 
DETOX, 
STUNT, 
DRILL, 
TUBER, 
GAUGE, 
PLUMP, 
FETID, 
HOLLY, 
GLARE, 
LATHE, 
ABHOR, 
MOTEL, 
RABID, 
AWOKE, 
VENUE, 
SHORT, 🟩🟩🟩🟩🟩
GUISE, 


In [10]:
# pre 1570 (PLANE)
subset_suspect_pre_1570 = subset_suspect[(subset_suspect['Game Number'] < 1570)]
subset_suspect_pre_1570

,Game Number,username,Answer,Output
24,1541,user_4,TENOR,🟨🟩⬜️⬜️⬜️
32,1542,user_4,CHIRP,🟨⬜️⬜️🟨⬜️
44,1543,user_4,TRICK,🟨⬜️⬜️🟨⬜️
46,1544,user_4,POUTY,🟩⬜️🟨🟨⬜️
55,1546,user_4,THROB,🟩🟩⬜️⬜️⬜️
60,1550,user_4,LEFTY,⬜️🟩⬜️🟩🟩
62,1551,user_4,TEETH,🟩🟩⬜️🟨⬜️
64,1552,user_4,KNIFE,⬜️🟨⬜️🟨⬜️
68,1553,user_4,LATER,🟨🟨⬜️⬜️🟨
71,1554,user_4,DEFER,🟩🟩⬜️⬜️⬜️


In [11]:
# post 1570 (PLANE) excluding 1600 (SHORT)
subset_suspect_post_1570 = subset_suspect[(subset_suspect['Game Number'] >= 1570) & (subset_suspect['Game Number'] != 1600)]
subset_suspect_post_1570

,Game Number,username,Answer,Output
111,1570,user_4,AMUSE,⬜️🟨⬜️⬜️🟨
115,1571,user_4,NYLON,⬜️⬜️⬜️⬜️⬜️
119,1572,user_4,ANNOY,⬜️⬜️⬜️⬜️🟨
123,1573,user_4,HARDY,🟨⬜️⬜️⬜️🟨
126,1574,user_4,LEVER,🟨🟩⬜️⬜️⬜️
129,1575,user_4,STACK,⬜️⬜️⬜️⬜️🟨
132,1576,user_4,WOUND,⬜️⬜️⬜️⬜️⬜️
135,1577,user_4,BEARD,🟨🟩⬜️⬜️🟨
138,1578,user_4,FORUM,🟨⬜️⬜️⬜️⬜️
141,1579,user_4,SPOOF,⬜️⬜️⬜️⬜️⬜️


In [12]:
# find the subset of a specific user
community_subset = csv[csv['username'] != suspect]
len(community_subset)

139

In [13]:
def mean_green(df):
    total_greens = 0
    total_entries = len(df)
    
    for output in df['Output']:
        total_greens += output.count('🟩')
    
    return total_greens / total_entries

This is the mean of how many green's are in the first guess
Ranging from 0 (if no greens were found) and 5 (if all guesses were perfect)
* $mean_{community} = 0.3381294964028777$
* $mean_{suspect} = 0.75$

Divide by 5 to have the range be from 0 to 1
* $\hat p_{community} = \frac{community_{mean}}{5} = 0.06762589928$
* $\hat p_{community} = \frac{community_{mean}}{5} = 0.15$

lengths of number of samples
* $n_{coummunity} = 139$
* $n_{suspect} = 56$

calculate the variance, multiplied by 5 because each sample has 5 letters
* $variance_{community}  = \frac{\hat p_{suspect} * (1 - \hat p_{suspect})}{(5 * c_{suspect})} = 0.00009072321874398417$
* $variance_{suspect}  = \frac{\hat p_{suspect} * (1 - \hat p_{suspect})}{(5 * c_{suspect})} = 0.0004553571428571429$

standard error difference
* $SE_{diff} = \sqrt{variance_{community} + variance_{suspect}} = 0.0233683624073474$

Z-score
* $z_score = \frac{\hat p_{community} - \hat p_{suspect}}{SE_{diff}} = 3.525026669970022$

In [14]:
# the mean can range from [0 to 5]
# divide by 5 to have the range be from [0 to 1]
def p_mean_hat(mean, divisor = 5):
    output = mean / divisor
    return output

In [15]:
def variance(n, mean, divisor = 5):
    p_hat_mean = p_mean_hat(mean, divisor)
    return p_hat_mean * (1 - p_hat_mean) / (n * divisor)

In [16]:
def standard_error_difference(variance_1, variance_2):
    return (variance_1 + variance_2) ** 0.5

In [17]:
def z_score(mean_1, mean_2, standard_error_difference):
    return (mean_1 - mean_2) / standard_error_difference

In [18]:
from scipy.stats import norm
def p_value_right_tail(z_score):
    return 1 - norm.cdf(z_score)

In [19]:
# number of first time guesses
n_community = len(community_subset)
n_suspect = len(subset_suspect)

In [20]:
mean_green_community = mean_green(community_subset)
mean_green_suspect= mean_green(subset_suspect)

In [21]:
# calculate p_mean_hat
p_mean_hat_community = p_mean_hat(mean_green_community)
p_mean_hat_suspect = p_mean_hat(mean_green_suspect)

In [22]:
# calculate the variance
variance_community = variance(n_community, mean_green_community)
variance_suspect = variance(n_suspect, mean_green_suspect)

In [23]:
# calculate the standard error difference
standard_error_difference_suspect = standard_error_difference(variance_community, variance_suspect)
standard_error_difference_suspect

0.0233683624073474

In [24]:
# calculate the z_score
z_score_suspect = z_score(p_mean_hat_suspect, p_mean_hat_community, standard_error_difference_suspect)
z_score_suspect

3.525026669970022

In [25]:
# calculate the p_value
p_value_right_tail_suspect = p_value_right_tail(z_score_suspect)
p_value_right_tail_suspect

0.00021172007892422684

The null hypothesis is that the suspect is playing fairly

In [26]:
def nullHypothesis(df_suspect, df_community, alpha = 0.01):
    print("Null hypothesis: The suspected user's mean number of green letters in the first guess, are equal to that of the average players.\n")
    
    n_suspect = len(df_suspect)
    n_community = len(df_community)
    mean_green_suspect = mean_green(df_suspect)
    mean_green_community = mean_green(df_community)
    print(f"Sample size (suspect): {n_suspect}")
    print(f"Sample size (community): {n_community}")
    print(f"Mean green letters (suspect): {mean_green_suspect:.3f}")
    print(f"Mean green letters (community): {mean_green_community:.3f}\n")
    
    
    p_hat_suspect = p_mean_hat(mean_green_suspect)
    p_hat_community = p_mean_hat(mean_green_community)
    variance_suspect = variance(n_suspect, mean_green_suspect)
    variance_community = variance(n_community, mean_green_community)
    se_diff = standard_error_difference(variance_community, variance_suspect)
    print(f"Estimated proportion (suspect): {p_hat_suspect:.4f}")
    print(f"Estimated proportion (community): {p_hat_community:.4f}")
    print(f"Variance (suspect): {variance_suspect:.6f}")
    print(f"Variance (community): {variance_community:.6f}")
    print(f"Standard error of difference: {se_diff:.6f}\n")

    z = z_score(p_hat_suspect, p_hat_community, se_diff)
    p_val = p_value_right_tail(z)
    print(f"Z-score: {z:.3f}")
    print(f"P-value (right-tailed): {p_val:.5f}")
    print(f"Significance level (α): {alpha}")
    
    if p_val > alpha:
        print("Fail to reject the null (no significant evidence of cheating).")
    else:
        print("Reject the null hypothesis (evidence suggests the user may be cheating).")
    

In [27]:
# before game 1570
nullHypothesis(subset_suspect, community_subset)

Null hypothesis: The suspected user's mean number of green letters in the first guess, are equal to that of the average players.

Sample size (suspect): 56
Sample size (community): 139
Mean green letters (suspect): 0.750
Mean green letters (community): 0.338

Estimated proportion (suspect): 0.1500
Estimated proportion (community): 0.0676
Variance (suspect): 0.000455
Variance (community): 0.000091
Standard error of difference: 0.023368

Z-score: 3.525
P-value (right-tailed): 0.00021
Significance level (α): 0.01
Reject the null hypothesis (evidence suggests the user may be cheating).


In [28]:
# after game 1570, excluding 1600
nullHypothesis(subset_suspect_pre_1570, community_subset)

Null hypothesis: The suspected user's mean number of green letters in the first guess, are equal to that of the average players.

Sample size (suspect): 24
Sample size (community): 139
Mean green letters (suspect): 1.208
Mean green letters (community): 0.338

Estimated proportion (suspect): 0.2417
Estimated proportion (community): 0.0676
Variance (suspect): 0.001527
Variance (community): 0.000091
Standard error of difference: 0.040223

Z-score: 4.327
P-value (right-tailed): 0.00001
Significance level (α): 0.01
Reject the null hypothesis (evidence suggests the user may be cheating).


In [29]:
# both combined
nullHypothesis(subset_suspect_post_1570, community_subset)

Null hypothesis: The suspected user's mean number of green letters in the first guess, are equal to that of the average players.

Sample size (suspect): 31
Sample size (community): 139
Mean green letters (suspect): 0.258
Mean green letters (community): 0.338

Estimated proportion (suspect): 0.0516
Estimated proportion (community): 0.0676
Variance (suspect): 0.000316
Variance (community): 0.000091
Standard error of difference: 0.020162

Z-score: -0.794
P-value (right-tailed): 0.78646
Significance level (α): 0.01
Fail to reject the null (no significant evidence of cheating).
